src/main/java/org/terasology/metalrenegades/combat/system/EnemySpawnSystem.java

```diff
@@ -0,0 +1,276 @@
+                spawnEnemyOnCharacter(character);
+            }
+        }
+
+        // Removes enemies that have entered a settlement
+        enemyQueue.removeIf(enemy -> {
+            LocationComponent locComp = enemy.getComponent(LocationComponent.class);
+            Vector3f enemyLoc = locComp.getWorldPosition();
+            if (settlementEntityManager.checkOutsideAllSettlements(new Vector2i(enemyLoc.getX(), enemyLoc.getZ()))) {
+                return false;
+            }
+            removeEnemy(enemy);
+            return true;
+        });
+
+        // If there are too many enemies in the world, remove the oldest enemies to make room.
+        while (enemyQueue.size() > MAX_ENEMIES) {
+            removeEnemy(enemyQueue.remove());
+        }
+    }
+
+    @ReceiveEvent
+    public void onDawnEvent(OnDawnEvent event, EntityRef entityRef) {
+        while (!enemyQueue.isEmpty()) {
+            removeEnemy(enemyQueue.remove());
+        }
+    }
+
+    @ReceiveEvent
+    public void onCharacterRespawn(OnPlayerRespawnedEvent event, EntityRef entity) {
+        entity.saveComponent(new EnemyGracePeriodComponent(5));
+    }
+
+    @ReceiveEvent(priority = EventPriority.PRIORITY_HIGH)
+    public void onEntityDestroyed(BeforeDestroyEvent event, EntityRef entityRef,
+                                  NightEnemyComponent nightEnemyComponent) {
+        entityRef.send(new RemoveCharacterFromOverlayEvent());
+        entityRef.destroy();
+        event.consume();
+    }
+
+    /**
+     * Spawns an enemy around the area of a particular character entity.
+     *
+     * @param character The spawn target character.
+     */
+    private void spawnEnemyOnCharacter(EntityRef character) {
+        LocationComponent locationComponent = character.getComponent(LocationComponent.class);
+
+        List<Vector3i> spawnPositions = findSpawnPositions(new Vector3i(locationComponent.getWorldPosition()));
+        Optional<Vector3i> potentialPos = spawnPositions.stream().findFirst();
+
+        if (potentialPos.isPresent()) {
+            spawnOnPosition(potentialPos.get());
+        }
+    }
+
+    /**
+     * Searches a 60x20x60 area around a given position for acceptable spawn positions.
+     *
+     * @param centrePos The position to search around.
+     * @return A list of possible spawn positions.
+     */
+    private List<Vector3i> findSpawnPositions(Vector3i centrePos) {
+        Vector3i worldPos = new Vector3i(centrePos);
+        List<Vector3i> foundPositions = Lists.newArrayList();
+        Vector3i blockPos = new Vector3i();
+        for (int y = -10; y < 10; y++) {
+            for (int z = -30; z < 30; z++) {
+                for (int x = -30; x < 30; x++) {
+                    blockPos.set(x + worldPos.x, y + worldPos.y, z + worldPos.z);
+                    if (isValidSpawnPosition.apply(blockPos)) {
+                        foundPositions.add(new Vector3i(blockPos));
+                    }
+                }
+            }
+        }
```

Comment:
The "check outside settlements" can be done at just the corners of the rectangle. Then for the actual spawn position pick a random point inside this rectangle, [raytrace](https://github.com/MovingBlocks/Terasology/blob/25b81f4a68b9ec5bd694cebc92c562c4e55a0fbf/engine/src/main/java/org/terasology/physics/Physics.java#L38) straight down to find the ground and spawn the enemy there. To further improve this, you could first pick 4 blocks (NSEW of the player) to see if they're near a water body or something that won't allow spawning. Then instead of picking a completely random point, have something like a weighted random that avoids invalid directions.

Reply:
Yep I agree. Could very well be simplified with finding a random point. You could just keep finding random positions inside this rectangle until you found a spawnable block. Raytracing is elegant, but it does introduce some form of complexity. You could just keep coming down block by block (from some reasonable height) until you find a non-penetrable block.

Reply:
The method of choosing spawn positions has been completely changed, using ideas from both responses. A list of valid chunks is collected, kept up to date with chunk respawn and despawn events. Then to choose a random spawn point, a random chunk is selected from this list, and a random (x, z) point is chosen in that chunk. The surface block in that column is found (by going down block by block, not with raytracing), and that's where the spawning occurs.

There are some cases where a column inside a chunk, or the entire chunk (for example, a chunk of all air) won't have any acceptable spawn points. In these cases, the spawn cycle will just be skipped. This can be swapped for trying again with a different chunk/column, but I want to avoid performance drops where spawn searching is attempted many times.